File: src/morphic/frontend/routes_resizer.py
"""
Resizer-tab API routes — scan, status, results.
"""

from __future__ import annotations

import os
import time

from flask import Blueprint, jsonify, request

from morphic.resizer.operations import RESIZE_MODES
from morphic.resizer.scanner import get_job, start_job

bp = Blueprint("resizer", __name__)


@bp.route("/scan", methods=["POST"])
def api_start_resize():
    """Start a batch resize job."""
    data = request.get_json()
    if not data:
        return jsonify({"error": "No JSON body"}), 400

    folder = data.get("folder", "").strip()
    width = data.get("width", 0)
    height = data.get("height", 0)
    mode = data.get("mode", "fit")
    bg_color = data.get("bg_color", "#000000")
    quality = data.get("quality", 90)
    output_folder = data.get("output_folder", "").strip() or None

    if not folder or not os.path.isdir(folder):
        return jsonify({"error": f"Invalid folder: {folder}"}), 400

    try:
        width = int(width)
        height = int(height)
    except (TypeError, ValueError):
        return jsonify({"error": "width and height must be integers"}), 400

    if width <= 0 or height <= 0:
        return jsonify({"error": "width and height must be positive"}), 400

    if mode not in RESIZE_MODES:
        return jsonify(
            {
                "error": f"mode must be one of {RESIZE_MODES}",
            }
        ), 400

    job_id = start_job(
        folder=folder,
        width=width,
        height=height,
        mode=mode,
        output_folder=output_folder,
        bg_color=bg_color,
        quality=quality,
    )
    return jsonify({"job_id": job_id}), 202


@bp.route("/scan/<job_id>/status")
def api_scan_status(job_id: str):
    """Get status of a resize job."""
    job = get_job(job_id)
    if not job:
        return jsonify({"error": "Job not found"}), 404

    elapsed = 0.0
    if job.started_at:
        end = job.finished_at if job.finished_at else time.time()
        elapsed = end - job.started_at

    return jsonify(
        {
            "id": job.id,
            "status": job.status,
            "progress": job.progress,
            "message": job.message,
            "error": job.error,
            "total_files": job.total_files,
            "processed_files": job.processed_files,
            "elapsed_seconds": round(elapsed, 1),
        }
    )


@bp.route("/scan/<job_id>/results")
def api_scan_results(job_id: str):
    """Get results of a completed resize job."""
    job = get_job(job_id)
    if not job:
        return jsonify({"error": "Job not found"}), 404

    if job.status not in ("done", "error"):
        return jsonify({"error": "Job not finished yet"}), 409

    return jsonify(
        {
            "results": job.results,
            "errors": job.errors,
            "total_files": job.total_files,
            "processed_files": job.processed_files,
        }
    )

File: web/routes_resizer.go
package web

import (
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/exterex/morphic/internal/resizer"
)

func registerResizerRoutes(r *gin.Engine) {
	g := r.Group("/api/resizer")
	{
		g.POST("/scan", handleResizerScan)
		g.GET("/scan/:id/status", handleResizerStatus)
		g.GET("/scan/:id/results", handleResizerResults)
	}
}

func handleResizerScan(c *gin.Context) {
	var req struct {
		Folder       string `json:"folder"`
		Width        int    `json:"width"`
		Height       int    `json:"height"`
		Mode         string `json:"mode"`
		BgColor      string `json:"bg_color"`
		Quality      int    `json:"quality"`
		OutputFolder string `json:"output_folder"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if req.Folder == "" || !isDir(req.Folder) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid folder: " + req.Folder})
		return
	}
	if req.Width <= 0 || req.Height <= 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "width and height must be positive"})
		return
	}
	if req.Mode == "" {
		req.Mode = "fit"
	}
	if !resizer.IsValidMode(req.Mode) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "mode must be one of fit, fill, stretch, pad"})
		return
	}
	if req.BgColor == "" {
		req.BgColor = "#000000"
	}
	if req.Quality == 0 {
		req.Quality = 90
	}

	jobID := resizer.StartJob(
		req.Folder, req.Width, req.Height, req.Mode,
		req.OutputFolder, req.BgColor, req.Quality,
	)
	c.JSON(http.StatusAccepted, gin.H{"job_id": jobID})
}

func handleResizerStatus(c *gin.Context) {
	id := c.Param("id")
	job, ok := resizer.GetJob(id)
	if !ok {
		c.JSON(http.StatusNotFound, gin.H{"error": "Job not found"})
		return
	}

	elapsed := 0.0
	if !job.StartedAt.IsZero() {
		end := job.DoneAt
		if end.IsZero() {
			end = time.Now()
		}
		elapsed = end.Sub(job.StartedAt).Seconds()
	}

	c.JSON(http.StatusOK, gin.H{
		"id":              job.ID,
		"status":          job.Status,
		"progress":        job.Progress,
		"message":         job.Message,
		"error":           job.Error,
		"total_files":     job.Total,
		"processed_files": job.Processed,
		"elapsed_seconds": round1(elapsed),
	})
}

func handleResizerResults(c *gin.Context) {
	id := c.Param("id")
	job, ok := resizer.GetJob(id)
	if !ok {
		c.JSON(http.StatusNotFound, gin.H{"error": "Job not found"})
		return
	}

	if job.Status != "done" && job.Status != "failed" {
		c.JSON(http.StatusConflict, gin.H{"error": "Job not finished yet"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"results":         job.Results,
		"errors":          job.Errors,
		"total_files":     job.Total,
		"processed_files": job.Processed,
	})
}
